buildSrc/src/main/java/com/google/firebase/gradle/ReleaseGenerator.kt

```diff
@@ -0,0 +1,118 @@
+        val currentRelease = project.property("currentRelease").toString()
+        val pastRelease = project.property("pastRelease").toString()
+        val rootDir = project.rootDir
+        val availableModules = parseSubProjects(rootDir)
+
+        val repo = initRepo(rootDir)
+        val headRef = repo.repository.resolve(Constants.HEAD)
+        val branchRef = getObjectRefForBranchName(repo, pastRelease)
+
+        val changedDirs = getChangedLibraries(repo, branchRef, headRef, FIREBASE_LIBRARIES)
+        val changedModules = findChangedModules(changedDirs, availableModules)
+        writeReleaseConfig(rootDir, changedModules, currentRelease)
+    }
+
+    private fun findChangedModules(changedDirs: List<String>, availableModules: Set<String>): List<String> = changedDirs.flatMap { directory ->
+        val moduleName = directory.replace("/", ":")
+        val ktxModuleName = "$moduleName:ktx"
+        listOf(moduleName, ktxModuleName).filter { availableModules.contains(it) }
+    }
+
+    private fun parseSubProjects(rootDir: File) = File(rootDir, "subprojects.cfg").readLines().filter { !it.startsWith("#") }.toSet()
+
+    @Throws(IOException::class)
+    private fun initRepo(repoPath: File): Git {
+        if (!repoPath.exists()) {
+            throw RuntimeException("Repo path doesn't exist")
+        }
+        return Git.open(repoPath)
+    }
+
+    @Throws(GitAPIException::class)
+    private fun getObjectRefForBranchName(repo: Git, branchName: String): ObjectId {
+        return repo.branchList().setListMode(ListBranchCommand.ListMode.REMOTE).call()
+                .filter { ref: Ref -> ref.name == "refs/remotes/origin/$branchName" }
+                .map { obj: Ref -> obj.objectId }.firstOrNull()
+                ?: throw RuntimeException("Could not find branch named $branchName")
+    }
+
+    private fun getChangedLibraries(repo: Git, startRef: ObjectId, endRef: ObjectId, directories: List<String>): List<String> {
+        return FIREBASE_LIBRARIES.filter { libraryPath: String ->
+            repo.log()
+                    .addPath("$libraryPath/*.kt")
+                    .addPath("$libraryPath/*.java")
+                    .addPath("$libraryPath/CHANGELOG.md")
+                    .addPath("$libraryPath/*.gradle")
+                    .addRange(startRef, endRef)
+                    .setMaxCount(1)
+                    .call()
+                    .iterator().hasNext()
+        }
+    }
+
+    private fun writeReleaseConfig(configPath: File, libraries: List<String>, releaseName: String) {
+        File(configPath, "release.cfg").writeText(
+                """
+                    [release]
+                    name = $releaseName
+                    mode = RELEASE
+                    
+                    [modules]
+                    ${libraries.joinToString("\n                    ")}
+                """.trimIndent()
+        )
+    }
+
+    companion object {
+        private val FIREBASE_LIBRARIES = listOf(
```

Comment:
Whats the difference between this and `subprojects.cfg`? Is this like a reduced list of who releases?

Reply:
+1. AFAIK all projects in `subprojects.cfg` do release.

Reply:
I had originally done this as a way to distinguish that -ktx libraries release alongside core libraries even if there are changes only in one of them. 

but I think i've come up with a better way to do this.